### PR TITLE
b/172826703: Ensure service-config related errors are actionable

### PR DESCRIPTION
**Pull request by nareddyt**
**Goal**: Ensures error messages are actionable. This requires providing extra context (JWT provider id, operation name, etc.) with a clear error message.

**Scope**: Let's focus on the Cloud API Gateway use-case first. The majority of user-facing errors for Cloud API Gateway will come from the service config (examples: bad JWT locations, invalid backend rule address, etc.) So this PR primarily focuses on errors related to processing the service config and creating envoy configs.

**Implementation details**:
- Majority of changes are only to error messages (no error handling behavior change).
- Some JWT errors are currently not handled. These can be fixed in a follow-up PR (b/176432170). This PR logs the errors and changes the method structure, but does not change error handling behavior.
- When calling `getOrCreateMethod` in `service_info.go`, always handle the error. This is a minor behavior change, it should be safe. Follow-up PR can cleanup usage of this method (b/176433373)

Signed-off-by: Teju Nareddy <nareddyt@google.com>

### CODE REVIEW

**Review by TAOXUY — APPROVED** [review]

File: src/go/configgenerator/cluster_generator.go [hunk 1]
```diff
@@ -122,7 +122,7 @@ func addDnsResolversToClusters(dnsResolverAddresses string, clusters []*clusterp
 func makeMetadataCluster(serviceInfo *sc.ServiceInfo) (*clusterpb.Cluster, error) {
 	scheme, hostname, port, _, err := util.ParseURI(serviceInfo.Options.MetadataURL)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("fail to parse metadata cluster URI: %v", err)
```
Comment (qiwzhang) [inline src/go/configgenerator/cluster_generator.go:125]: Should we also print the problem URL for easy debug?

**Review by qiwzhang — CHANGES_REQUESTED** [review]
In most cases, we only print error messages, I think we should print error url too

File: src/go/util/url_util.go [hunk 2]
```diff
@@ -150,7 +150,7 @@ func IamAccessTokenPath(IamServiceAccount string) string {
 	return fmt.Sprintf("/v1/projects/-/serviceAccounts/%s:generateAccessToken", IamServiceAccount)
 }
 
-func ExtraAddressFromURI(jwksUri string) (string, error) {
+func ExtractAddressFromURI(jwksUri string) (string, error) {
```
Comment (qiwzhang) [inline src/go/util/url_util.go:153]: s/jwksUri/uri

Comment (nareddyt) [reply src/go/util/url_util.go:153]: Done

File: src/go/configinfo/service_info.go [hunk 3]
```diff
@@ -244,14 +251,18 @@ func (s *ServiceInfo) addGrpcHttpRules() error {
 	for _, api := range s.serviceConfig.GetApis() {
 		for _, method := range api.GetMethods() {
 			selector := fmt.Sprintf("%s.%s", api.GetName(), method.GetName())
-			mi, _ := s.getOrCreateMethod(selector)
+			mi, err := s.getOrCreateMethod(selector)
+			if err != nil {
+				return fmt.Errorf("error processing auto-generated gRPC http rule for operation (%s.%s): %v", api.GetName(), method.GetName(), err)
```
[the remark this answers is not in the document]
Comment (nareddyt) [reply src/go/configinfo/service_info.go]: Rephrased. I would still like to mention we are creating a gRPC http rule, that is the main point of this method.

File: src/go/configinfo/service_info.go [hunk 4]
```diff
@@ -210,13 +213,16 @@ func (s *ServiceInfo) processEmptyJwksUriByOpenID() error {
...
+func (s *ServiceInfo) processApis() error {
 	for _, api := range s.serviceConfig.GetApis() {
 		s.ApiNames = append(s.ApiNames, api.Name)
 
 		for _, method := range api.GetMethods() {
 			selector := fmt.Sprintf("%s.%s", api.GetName(), method.GetName())
-			mi, _ := s.getOrCreateMethod(selector)
+			mi, err := s.getOrCreateMethod(selector)
+			if err != nil {
+				return fmt.Errorf("error processing operation (%v): %v", selector, err)
```
[the remark this answers is not in the document]
Comment (nareddyt) [reply src/go/configinfo/service_info.go]: Done. Let's not worry too much about the error messages for `getOrCreateMethod(...)`. I will work on b/176433373 next to cleanup usage of this.

File: src/go/configgenerator/cluster_generator.go (hunk 1 above)
Comment (nareddyt) [reply src/go/configgenerator/cluster_generator.go:125]: Done. I added this logic to the shared `ParseURI` function to reduce code duplication.

**Review by qiwzhang — APPROVED** [review]

**Pull request merged**